### DISCUSSION TITLE
Render button doesn't work when raw work done in R script; render("xyz.qmd") in command line does.

### DISCUSSION THREAD

**Discussion by mm225022** [quarto-dev/quarto-cli#3958] [Q&A] [answered]
I do my raw work in an R script.  Consequently, all of my libraries and variables are loaded into the enviroment.  I write my RMarkdown/Quarto document, synthesize the analysis and produce the report, referencing those objects that I have created.

I still do some work in the RMarkdown/Quarto document.  I never had any problem with this approach in RMarkdown.  However, when I attempt it in Quarto, and I press the render button, it errors out.  The most recent time I did this, I got a pop-up that stated:

`Error when opening file.   The system cannot find the file specified.`

In the terminal window (under Background Jobs), the error message displayed in this specific example is:

```
Error in full_join(cbind(ABC = "abcd", xyz[2, 2:ncol(xyz)]/xyz[3, 2:ncol(xyz)]) %>%  : 
  could not find function "%>%"
Calls: .main ... withVisible -> eval_with_user_handlers -> eval -> eval

Execution halted
```

In this case, I am not sure why it is stating it cannot find `%>%`.  I have `library(dplyr)` clearly loaded for the work that I am doing with my R script (which has already been done within this environment).

However, I am still able to render my Quarto file the old RMarkdown way by typing render("Quarto_Doc.qmd") in my R script window.

Does anyone have a similar issue and can anyone advise on how to get around?

Thank

**Comment by mcanouil** [comment, marked as answer]
Quarto is not executed in your current R session. Quarto, if needed, creates its own R session to compute the R code cells within the document.

RMarkdown is fully embedded into R.

RMarkdown is not supposed to use the global environment of an active R session.

Note that when you use knitr/rmarkdown to render a Quarto document, you are not using Quarto at all.

FYI, code cells/chunks of RMardown and Quarto documents can be executed in RStudio or VSCode straight from them.

**Reply by cderv** [reply]
Thanks @mcanouil ! 

> RMarkdown is not supposed to use the global environment of an active R session.

About this @jonathanmburns , R Markdown will use by default the R session in which `rmarkdown::render()` is called. So if you have taken the habit of doing stuff in your current working then calling `rmarkdown::render()` in console or everything in one R script, then yes it works. 

However, if you would click on the Knit button in rstudio IDE for example, your document would be sent in background session to render, so  it would not use current global environment. 

This is why good practice for reproducibility (so that your R Markdown file can be rendered not matter how) is that every library and object is to be loaded from within a chunk in the document. If you were to sent your Rmd file to someone else, one would try to render the file in a clean session and it is expected to work. 

With Quarto, rendering of a qmd with R chunks will open a specific R session in  background with no prior knowledge of another environment. So this practice for reproducibiliy becomes the only practice with Quarto : Everything must be defined, loaded and done inside a single Qmd document in a chunk. You can use **knitr** and any R function to source an external script, load object or else but it must be done in the `.qmd` file. That is by design. 

hope it helps understand with more context the different.

**Reply by mm225022** [reply]
Thanks very much for the clarification @cderv .  My problem (and thought process) has always been that I do my EDA and rough work in R session.  Once I've finalized the code and the specific things that I want to include the report, then I build the RMarkdown file.  I wouldn't want to include all of those background code chunks in the RMarkdown file as it would get quite messy.  Also, while you can visual your results in RMarkdown, I personally find it much cleaner to do so (at least for the EDA and the like) in the "regular" R session.
It sounds like you are saying that the simple solution is just to source that R script file with all the background work and objects at the beginning of my Quarto document, and then I'll be away to the races.  I am understanding correctly?

Thanks again

**Reply by cderv** [reply]
> It sounds like you are saying that the simple solution is just to source that R script file with all the background work and objects at the beginning of my Quarto document, and then I'll be away to the races. I am understanding correctly?

Yes you can source a R script in the first chunk and hide the output using `include: false` - This way your R code would be evaluated in the R session for the rendering and the remaining chunks would have access to all created objects by this R script. 

* Source an external script : https://bookdown.org/yihui/rmarkdown-cookbook/source-script.html
* Load a script as a chunk code content : https://bookdown.org/yihui/rmarkdown-cookbook/option-code.html

Both would do the same effect, only the latter would show the actual R code from the script inside the code chunk in output. 

This would work with Quarto as **knitr** is used also.
